### PR TITLE
Add periodic health monitoring for vMCP backends

### PR DESCRIPTION
**Pull request by yrobla**
Implement health check system that monitors backend MCP server availability through periodic ListCapabilities calls. This is the foundation for the health monitoring and circuit breaker system described in issue #3036.

Addresses first part of #3036

## Large PR Justification

This is an atomic feature, including tests to validate that the feature is functionally complete.

### CODE REVIEW

File: cmd/thv-operator/controllers/virtualmcpserver_controller.go [hunk 1]
```diff
@@ -1656,19 +1712,188 @@ func (r *VirtualMCPServerReconciler) discoverBackends(
...
+	default:
+		// Default to unknown for unexpected status values
+		return mcpv1alpha1.BackendStatusUnknown
+	}
+}
+
+// phaseChangePredicate triggers reconciliation when status.phase changes on backend resources.
+// This ensures VirtualMCPServer reconciles when backends transition between Running/Pending/Failed states.
+// Works with both MCPServer and MCPRemoteProxy resources.
+type phaseChangePredicate struct {
```
Comment (jhrozek) [inline cmd/thv-operator/controllers/virtualmcpserver_controller.go:1840]: **HIGH: Incomplete phaseChangePredicate Implementation**

The `phaseChangePredicate` embeds `predicate.Funcs` but only overrides `Update()`. The default implementations for `Create()`, `Delete()`, and `Generic()` return `false`, which may cause the controller to miss important events:

- **Create**: New MCPServers/MCPRemoteProxies won't trigger reconciliation until an update occurs
- **Delete**: Deleted backends won't be detected until the next scheduled reconciliation

This could cause stale health status in the VirtualMCPServer CRD.

```suggestion
type phaseChangePredicate struct {
	predicate.Funcs
}

// Create returns true to ensure new resources trigger reconciliation
func (phaseChangePredicate) Create(_ event.CreateEvent) bool {
	return true
}

// Delete returns true to ensure deleted resources trigger reconciliation
func (phaseChangePredicate) Delete(_ event.DeleteEvent) bool {
	return true
}
```

Then add the `Update` method below these.

File: cmd/thv-operator/controllers/virtualmcpserver_controller.go [hunk 2]
```diff
@@ -2096,3 +2321,123 @@ func (*VirtualMCPServerReconciler) vmcpReferencesCompositeToolDefinition(
...
+	}
+	if timeout < minHealthCheckQueryTimeout {
+		timeout = minHealthCheckQueryTimeout
+	}
+
+	// Create HTTP client with derived timeout
+	// Note: Uses default transport which validates TLS certificates.
+	// If the vmcp server uses self-signed certificates, ensure proper cert configuration
+	// (e.g., via cert-manager) or use HTTP for internal cluster communication.
+	httpClient := &http.Client{
```
Comment (jhrozek) [inline cmd/thv-operator/controllers/virtualmcpserver_controller.go:2382]: **MEDIUM: HTTP Client Created Per Request**

A new `http.Client` is created on every call to `queryVMCPHealthStatus`. While the health status cache (good addition!) reduces call frequency, this pattern still:
- Creates unnecessary GC pressure
- Prevents connection pooling and keep-alive benefits
- Is inconsistent with Go best practices (`http.Client` is safe for concurrent use)

Consider creating a shared `http.Client` at controller initialization:

```go
type VirtualMCPServerReconciler struct {
    // ... existing fields ...
    httpClient *http.Client // Add this
}

// In SetupWithManager:
func (r *VirtualMCPServerReconciler) SetupWithManager(mgr ctrl.Manager) error {
    r.httpClient = &http.Client{
        Timeout: defaultHealthCheckTimeout,
        Transport: &http.Transport{
            MaxIdleConns:        100,
            MaxIdleConnsPerHost: 10,
            IdleConnTimeout:     90 * time.Second,
        },
    }
    // ...
}
```

Note: Adjust the per-request timeout using context timeouts instead.

File: cmd/thv-operator/controllers/virtualmcpserver_controller.go [hunk 3]
```diff
@@ -2096,3 +2321,123 @@ func (*VirtualMCPServerReconciler) vmcpReferencesCompositeToolDefinition(
...
+			r.healthStatusCacheMutex.RUnlock()
+			ctxLogger.V(1).Info("Using cached health status", "url", healthURL, "expires_in", time.Until(cached.expiresAt))
+			return cached.healthStatus
+		}
+	}
+	r.healthStatusCacheMutex.RUnlock()
+
+	// Derive timeout from health check interval
+	// Use the health check interval or default, with a minimum to ensure reasonable timeout
+	timeout := healthCheckInterval
```
Comment (jhrozek) [inline cmd/thv-operator/controllers/virtualmcpserver_controller.go:2370]: **MEDIUM: No Upper Bound on HTTP Timeout**

The timeout is derived directly from `healthCheckInterval` with only a minimum bound:

```go
timeout := healthCheckInterval
if timeout < minHealthCheckQueryTimeout {
    timeout = minHealthCheckQueryTimeout
}
```

If a user configures `HealthCheckInterval: "10m"`, the HTTP client will have a 10-minute timeout. Long-running HTTP requests during reconciliation can:
- Block the reconciliation queue (Kubernetes has `MaxConcurrentReconciles`, but one slow request affects concurrency)
- Exhaust controller resources

Consider adding a maximum bound:

```go
// Cap at reasonable maximum to prevent blocking reconciliation
const maxHealthCheckQueryTimeout = 30 * time.Second
if timeout > maxHealthCheckQueryTimeout {
    timeout = maxHealthCheckQueryTimeout
}
```

File: pkg/vmcp/health/checker.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+		return vmcp.BackendUnhealthy
+	}
+
+	// Default to unhealthy for unknown errors
+	return vmcp.BackendUnhealthy
+}
+
+// isAuthError checks if the error message indicates an authentication failure.
+// Uses more specific patterns to avoid false positives from substrings in hostnames, URLs, etc.
+func isAuthError(errMsg string) bool {
```
Comment (jhrozek) [inline pkg/vmcp/health/checker.go:108]: **MEDIUM: Fragile String-Based Error Categorization**

The `isAuthError`, `isTimeoutError`, and `isConnectionError` functions use string matching on error messages. This is fragile because:
- Error messages can change between Go versions or library updates
- Localized error messages may not match
- Substring matching can produce false positives (e.g., a hostname containing "401")

The codebase already has typed errors in `pkg/vmcp/errors.go`:
- `ErrAuthenticationFailed`
- `ErrTimeout`  
- `ErrBackendUnavailable`

Consider using `errors.Is()` or `errors.As()` for type-safe error categorization:

```go
func categorizeError(err error) vmcp.BackendHealthStatus {
    if err == nil {
        return vmcp.BackendHealthy
    }
    if errors.Is(err, vmcp.ErrAuthenticationFailed) {
        return vmcp.BackendUnauthenticated
    }
    if errors.Is(err, vmcp.ErrTimeout) || errors.Is(err, context.DeadlineExceeded) {
        return vmcp.BackendUnhealthy
    }
    // Fallback to string matching for external errors
    // ...
}
```

This would require wrapping errors appropriately in the client code, but provides a more robust solution.

File: cmd/thv-operator/controllers/virtualmcpserver_controller.go [hunk 5]
```diff
@@ -2096,3 +2321,123 @@ func (*VirtualMCPServerReconciler) vmcpReferencesCompositeToolDefinition(
...
+		BackendID           string    `json:"backendId"`
+		Status              string    `json:"status"`
+		ConsecutiveFailures int       `json:"consecutiveFailures"`
+		LastCheckTime       time.Time `json:"lastCheckTime"`
+		LastError           string    `json:"lastError,omitempty"`
+		LastTransitionTime  time.Time `json:"lastTransitionTime"`
+	} `json:"backends"`
+}
+
+// queryVMCPHealthStatus queries the vmcp health endpoint and returns backend health information.
```
Comment (jhrozek) [inline cmd/thv-operator/controllers/virtualmcpserver_controller.go:2343]: **MEDIUM: Missing Unit Tests for Controller Health Functions**

The new controller functions lack dedicated unit tests:
- `queryVMCPHealthStatus` - HTTP mocking would validate response parsing
- `determineStatusFromBackends` - only 1 reference in test file vs multiple new code paths
- `mapHealthStatusToCRDStatus` - simple but should have table-driven tests
- `isPhaseUnhealthy` - should verify all phase combinations

Consider adding tests to `virtualmcpserver_controller_test.go` or a dedicated `virtualmcpserver_health_test.go`.

File: pkg/vmcp/health/checker.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+		logger.Debugf("Health check failed for backend %s: %v (status: %s)",
+			target.WorkloadName, err, status)
+		return status, fmt.Errorf("health check failed: %w", err)
+	}
+
+	logger.Debugf("Health check succeeded for backend %s", target.WorkloadName)
+	return vmcp.BackendHealthy, nil
+}
+
+// categorizeError determines the appropriate health status based on the error type.
```
Comment (jhrozek) [inline pkg/vmcp/health/checker.go:82]: **MEDIUM: Missing Test Coverage for checker.go**

While `monitor_test.go` exists, there's no `checker_test.go` covering:
- `categorizeError` function with various error types
- `isAuthError`, `isTimeoutError`, `isConnectionError` edge cases
- `CheckHealth` with mocked client

Table-driven tests would be valuable here given the string-matching complexity.

File: pkg/vmcp/server/server.go [hunk 7]
```diff
@@ -582,6 +666,69 @@ func (*Server) handleHealth(w http.ResponseWriter, _ *http.Request) {
...
+	for backendID, state := range states {
+		response := BackendHealthResponse{
+			BackendID:           backendID,
+			Status:              string(state.Status),
+			ConsecutiveFailures: state.ConsecutiveFailures,
+			LastCheckTime:       state.LastCheckTime,
+			LastTransitionTime:  state.LastTransitionTime,
+		}
+		if state.LastError != nil {
+			response.LastError = state.LastError.Error()
```
Comment (jhrozek) [inline pkg/vmcp/server/server.go:717]: **LOW: Error Messages May Expose Internal Information**

The `LastError` field exposes raw error messages via the health API:

```go
response.LastError = state.LastError.Error()
```

While the endpoint is internal to the cluster, error messages may contain:
- Internal hostnames/IPs
- File paths
- Stack traces (depending on error wrapping)

Consider sanitizing or categorizing errors before exposure:

```go
if state.LastError != nil {
    // Return category instead of full error
    response.LastError = categorizeErrorForDisplay(state.LastError)
}
```

File: cmd/thv-operator/controllers/virtualmcpserver_controller.go [hunk 8]
```diff
@@ -2096,3 +2321,123 @@ func (*VirtualMCPServerReconciler) vmcpReferencesCompositeToolDefinition(
...
+	if resp.StatusCode != http.StatusOK {
+		ctxLogger.Info("Unexpected status code from vmcp health endpoint",
+			"url", healthURL,
+			"status_code", resp.StatusCode)
+		return nil
+	}
+
+	// Parse response
+	var healthResp BackendHealthStatusResponse
+	if err := json.NewDecoder(resp.Body).Decode(&healthResp); err != nil {
```
Comment (jhrozek) [inline cmd/thv-operator/controllers/virtualmcpserver_controller.go:2416]: **LOW: Missing Response Body Size Limit**

The JSON decoder reads directly from the response body without size limits:

```go
if err := json.NewDecoder(resp.Body).Decode(&healthResp); err != nil {
```

While this is an internal endpoint, a malicious or misconfigured vMCP pod could return a very large response causing memory issues.

Consider using `io.LimitReader`:

```go
// Parse response with size limit to prevent memory exhaustion
const maxResponseSize = 1 << 20 // 1MB should be plenty for health status
limitedBody := io.LimitReader(resp.Body, maxResponseSize)
var healthResp BackendHealthStatusResponse
if err := json.NewDecoder(limitedBody).Decode(&healthResp); err != nil {
```
